---
marp: true
title: Representing biological sequences
author: Cheng Soon Ong
theme: gaia
class: invert
---

<style>
.container{
    display: flex;
}
.colbig{
    flex: 70%;
}
.colsmall{
    flex: 30%;
}
.col{
    flex: 1;
}
.highlight{
    color: lightcoral;
}
.cite{
    color: dodgerblue;
}
.footnote{
    color: dodgerblue;
    font-size: 70%;
}
.dna{
    color: lightseagreen;
    font-family: "Times New Roman", Times, serif;
    font-weight: bold;
    font-size: 120%;
}
</style>

<!-- _class: lead -->

# Opportunities and challenges in the sequencing revolution

<span class="highlight">Cheng Soon</span> Ong

---

The bottleneck in genome sequencing is no longer data generation 
-- the computational challenges around data analysis, display and integration are now rate limiting.

<div class="highlight">
    New approaches and methods are required to meet these challenges.
</div>

![bg right:20% 80%](figs-bio/genomics-computation-challenge.jpg)

<div class="cite">
    Green, Guyer and National Human Genome Research Institute
    Charting a course for genomic medicine from base pairs to bedside, Nature 2011.
</div>


---

<!-- _class: lead -->

# A glimpse of molecular biology

---
# Central dogma


- **DNA** written 5' to 3'.
  e.g. <span class="dna">AATCGAAGTTA</span>
- **RNA** T $\Rightarrow$ U
    e.g. <span class="dna">AAUCGAAGUUA</span>
- **Amino acid** 
  - 3 letters of RNA (codon) $\Rightarrow$ amino acid,
  - 20 letter alphabet.

<span class="cite">Lewin, Genes</span>


![bg right:50% 100%](figs-bio/central-dogma.png)

---
<!-- _class: lead -->

# Gene finding

---

# Genes are pure information objects

“There can be little doubt,” philosopher and biochemist Lenny Moss claimed in 2003, “that the idea of ‘the gene’ has been the central organizing theme of twentieth century biology”. And yet it is clear that the science of genetics never provided one generally accepted definition of the gene.
https://plato.stanford.edu/entries/gene/


---

# Gene finding

#### Given the DNA, predict resulting mRNA and protein

- Requires very accurate identification of
    - splice sites, translation \& transcription starts \& stops
    - sites of regulation (transcription, splicing, etc.)
- Develop methods to integrate single site predictions
    - usually HMMs and recurrent models
    - Novel learning methods for structured outputs


---

# Transcription start

![height:200px](figs-bio/promotor6.png)

- POL II binds to a rather vague region of  $\approx [-20,+20]$ bp
- Upstream of  TSS: promoter, with transcription factor binding sites
- Downstream of TSS: 5' UTR, and further downstream coding regions and introns (different statistics)
- 3D structure allows the transcription factors to bind


---

# Transcription termination

![height:200px](figs-bio/polyAcleavage.png)

- Polyadenylation signal (AATAAA or variants) 10-30 bp upstream
- T-rich or GT-rich elements 20-40 bp downstream
- Transcription end is several hundreds of bp after 3' cleavage site, mechanism poorly understood

---

# Splice sites (I)

- **Finding Intron-Exon junctions**
    ![width:800px](figs-bio/splice-window.png)
    known splice site
    - <span class="highlight">true sites</span>: fixed-length window around splice site
    - <span class="highlight">decoys sites</span>: generated by shifting the window

---

# Splice sites (II)

- **Finding Intron-Exon junctions**
    ![width:800px](figs-bio/splice-window2.png)
    $\Rightarrow$ Very unbalanced problem (1:200)
    $\Rightarrow$ Billions of data points from sequence databases
    $\Rightarrow$ Large scale methods necessary



---

# Splice forms

![bg right:60% 100%](figs-bio/gene-finding.png)

Predict a sequence of binary decisions
<span class="cite">Schweikert et. al. mGene: accurate SVM-based gene finding with an application to nematode genomes, 2009</span>

---

# Alternative splicing

![height:250px](figs-bio/alternative-splicing-graph.png)

- Nematodes and humans have roughly the same number of genes.
- C. elegans has 959 cells.

<span class="cite">Kahles et. al., SplAdder: identification, quantification and testing of alternative splicing events from RNA-Seq data, 2016</span>

---

# Regulation and control

<div class="container">

<div class="colbig">

- Genes are regulated by proteins called transcription factors (and promoters, enhancers, ribosomal binding site, etc.)
- Environment, e.g. metabolism (internal), temperature (external)

</div>
<div class="colsmall">

![](figs-bio/Gene_Regulatory_Network.jpg)

</div>
</div>

<div class="cite">

Alon, An Introduction to Systems Biology, 2007
Lawrence et. al. Learning and Inference in Computational Systems Biology, 2010

</div>

---

# Chromatin structure

![bg right:30% 100%](figs-bio/DNA_to_Chromatin_Formation.jpg)

- DNA packed tightly in nucleus
- DNA wrapped around histones to form nucleosomes
- Nucleosomes organised into chromatin
- Transcription accessibility
- DNA repair

https://www.nature.com/scitable/topicpage/dna-packaging-nucleosomes-and-chromatin-310/

---

# Methylation and Epigenetics


![bg right:40% 100%](figs-bio/epigenetics.png)

- Environmental effects change the genome (and may be inherited)
- Can detect changes using sequencing methods

<div class="cite">

Aristizabal et. al. A primer on epigenetics, 2020
Jonkhout et. al. The RNA modification landscape in human disease, 2017.

</div>

---

# Glimpse of molecular biology

- **DNA** Positive strand, written 5' to 3'.
      e.g. AATCGAAGTTA
- **RNA** T $\Rightarrow$ U
      e.g. AAUCGAAGUUA
- **Amino acid** 3 letters of RNA (codon) $\Rightarrow$ amino acid
- **Splicing** pre-mRNA to mature mRNA
- **Transcription factor** Regulate expression of gene, 
    through promoters and repressors
- **Epigenetics** Methylation, Chromatin marks


---
<!-- _class: lead -->

# The sequencing revolution

---

# History of sequencing

-  1960s: DNA - properties, proto sequencing
-  70s-90s: Manual sequencing - Sanger, Maxam-Gilbert
-  90s: Automated Sanger - flourescent, clones, colony picking
-  2003: Human genome - 25 cents per 1000 bases
-  00s: NGS, Clusters - 454-Roche, Solexa-Illumina, Ion Torrent
-  Illumina HiSeq X Ten: 6 billion 150 base sequences in 3 days

http://phylogenomics.blogspot.com.au/2015/10/evolution-of-dna-sequencing-talk-2015.html

---

# $1000 human genome

<div class="container">

<div class="col">

- **Data volume**
    -  HiSeq X Ten: 12 GB per hour
    -  700MB per human genome $\sim$ 200GB reads
- **Multiple versions**
    -  Multiplexing - tag sequences
    -  Small amounts of DNA
    -  Longer reads

</div>
<div class="col">

- **Small sequencers**
    -  Single cell sequencing: PacBio
    -  Real time sequencing: Oxford Nanopore

</div>
</div>

---

<!-- _class: lead -->
DNA sequences: 6 billion reads, each 150 bases, ~$\frac{1}{2}$TB

# So what?

---

# We have a new cheap sensor

![height:280px](figs-bio/illumina-hiseq.png) &emsp; ![height:280px](figs-bio/minion-sequencing.jpg)

---

# DNA sequencing

- **Analogy: Shotgun sequencing**
    Take many copies of a text, split at random points, reconstruct.
- **Alignment**
    - Dynamic programming
    - Needleman-Wunsch and Smith-Waterman
- **Assembly**
    - reference genome vs de-novo vs pan-genome
    - Bridges of K\"onigsberg $\rightarrow$ de Bruijn graphs
    http://www.nature.com/nmeth/journal/v9/n4/full/nmeth.1935.html


---

# Cohort (I)

- **Single Nucleotide Variation**
    -  Recall two copies of chromosomes
    -  at every location: AA, AB, BB
    -  Noise free, high coverage $\Rightarrow$ frequency = probability
    -  Probabilistic methods for maximum a posteriori estimation
    -  Correlations along the genome

1000 genomes project
https://www.internationalgenome.org/

---

# Cohort (II)

- **Structural variation**
    -  copy number variation
    -  insertions, deletions
    -  inversion, translocation
    <span class="cite">Cmero et. al. Inferring structural variant cancer cell fraction, 2020</span>

- **Study cohort** 
    - germline vs somatic mutations
    https://www.ncbi.nlm.nih.gov/books/NBK21894/

---

## Replication
#### Multiple samples to estimate experimental variation
- **Techical**
    -  Effect of measurement instrument
    -  Different days, researcher
    -  Usually same biological sample
- **Biological**
    -  Effect of biological development
    -  Different individuals of same ``species''

https://xkcd.com/2533/

---

# RNA-seq

- **Chemistry** 
    - Convert RNA to DNA
- **Gene Expression**
    -  Recall: mRNA translated to proteins
    -  Which genes are expressed in what tissues at which levels?
    -  What are the regulators of a particular gene?
    -  How does treatment change expression (differential expression)?
    https://www.encodeproject.org/


---

<div class="container">

<div class="colsmall">

# *-seq

- RNA-seq
- Hi-C-seq
- Sort-seq
- ChIP-seq
- scRNA-seq
$$\vdots$$
- hundreds more

</div>

<div class="colbig">

![](figs-bio/seqcartoon.jpg)
https://liorpachter.wordpress.com/seq/
<span class="cite">Image from Lior Pachter's ISMB 2013 keynote</span>

</div>
</div>

